src/fs/metadata_log.hh

```diff
@@ -0,0 +1,395 @@
+            // This order is linear and we always lock in one direction, so the graph of locking relations (A -> B iff
+            // lock on A is acquired and lock on B is acquired / being acquired) makes a DAG. Thus, deadlock is
+            // impossible, as it would require a cycle to appear.
+            std::pair<inode_t, std::optional<std::string>&> k1 {kind1.inode, kind1.dir_entry};
+            std::pair<inode_t, std::optional<std::string>&> k2 {kind2.inode, kind2.dir_entry};
+            if (k1 < k2) {
+                return with_locks_in_order(std::move(kind1), std::move(kind2), std::forward<Func>(func));
+            } else {
+                return with_locks_in_order(std::move(kind2), std::move(kind1), std::forward<Func>(func));
+            }
+        }
+    } _locks;
+
+    // TODO: for compaction: keep some set(?) of inode_data_vec, so that we can keep track of clusters that have lowest
+    //       utilization (up-to-date data)
+    // TODO: for compaction: keep estimated metadata log size (that would take when written to disk) and
+    //       the real size of metadata log taken on disk to allow for detecting when compaction
+
+    friend class metadata_log_bootstrap;
+    friend class create_file_operation;
+    friend class close_file_operation;
+    friend class unlink_or_remove_file_operation;
+
+public:
+    metadata_log(block_device device, unit_size_t cluster_size, unit_size_t alignment,
+            shared_ptr<metadata_to_disk_buffer> cluster_buff);
+
+    metadata_log(const metadata_log&) = delete;
+    metadata_log& operator=(const metadata_log&) = delete;
+
+    future<> bootstrap(inode_t root_dir, cluster_id_t first_metadata_cluster_id, cluster_range available_clusters,
+            fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id);
+
+private:
+    bool inode_exists(inode_t inode) const noexcept {
+        return _inodes.count(inode) != 0;
+    }
+
+    void write_update(inode_info::file& file, inode_data_vec new_data_vec);
+
+    // Deletes data vectors that are subset of @p data_range and cuts overlapping data vectors to make them not overlap
+    void cut_out_data_range(inode_info::file& file, file_range range);
+
+    void memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata);
+    void memory_only_update_metadata(inode_t inode, unix_metadata metadata);
+    void memory_only_delete_inode(inode_t inode);
+    void memory_only_small_write(inode_t inode, disk_offset_t offset, temporary_buffer<uint8_t> data);
+    void memory_only_update_mtime(inode_t inode, decltype(unix_metadata::mtime_ns) mtime_ns);
+    void memory_only_truncate(inode_t inode, disk_offset_t size);
+    void memory_only_add_dir_entry(inode_info::directory& dir, inode_t entry_inode, std::string entry_name);
+    void memory_only_delete_dir_entry(inode_info::directory& dir, std::string entry_name);
+
+    template<class Func>
+    void schedule_background_task(Func&& task) {
```

Comment:
This queue of background tasks is unbounded, which is worrying - effectively we can have an unbounded parallelism for background operations, and it's easy to overload/OOM the server that way.

This one would be better if it was:
1. Coordinated by an external "background worker" service
2. Throttled

This service can simply be a seastar thread which has a big loop which is guarded with a parallelism-bounding semaphore and simply takes stuff from a queue and runs it. To provide a pseudo-seastar example:

```cpp
class background_worker {
    static constexpr int max_parallelism = 1000;
    semaphore _sem{max_parallelism};
    some_kind_of_queue _background_tasks_to_run;
    abort_source _as;
    future<> _run_worker_thread;
public:
    background_worker () : _run_worker_thread(start_worker()) {}

    future<> stop() {
        _as.request_abort();
       _sem.broken();
        return std::move(_started);
    }

    future<> enqueue_task(some_task_type&& task) {
        return _background_tasks_to_run.push(std::move(task));
    }

protected:
    future<> start_worker() {
        return async [] {
            while (!_as.abort_requested()) {
                wait_for_the_semaphore_to_reduce_parallelism().get();
                wait_until_the_queue_is_not_empty().get();
                auto task_to_run = _background_tasks_to_run.pop_front();
                task_to_run().get();
            }
        }
    }
}
```

and then, you just start the background worker as part of the file system runtime for each shard, and then `schedule_background_task` can be implemented as:
```cpp
future<> schedule_background_task(Func&& task) {
    return _background_worker.enqueue_task(std::move(task));
}
```
Note that `schedule_background_task` returns a future, which reduces the unbounded parallelism of background tasks.